src/gravel/controllers/kv.py

```diff
@@ -12,125 +12,370 @@
         """Open k/v store connection"""
-        # try getting the status, loop until we make it.
-        opened = False
-        while not self._is_closing:
+        # nothing to do here, the connection thread is started
+        # automatically in __init__(), but I do kinda wonder if we
+        # should wait here a bit, at least, to give it time to
+        # connect initially, assuming the cluster exists...?
+        await asyncio.sleep(2)
+
+
+    def _cluster_connect(self):
+        logger.debug("Starting cluster connection thread")
+        while self._run:
             try:
-                async with aetcd3.client() as client:
-                    await client.status()
-            except Exception:
-                logger.warn("etcd not up yet? sleep.")
-                await asyncio.sleep(1.0)
-                continue
-            opened = True
-            break
-        if opened:
-            self._client = aetcd3.client()
-            self._is_open = True
-            logger.info("opened kvstore connection")
+                if not self._cluster:
+                    # uses /etc/ceph/ceph.client.admin.keyring
+                    # really should do separate keys per node so they can be
+                    # evicted if necessary if nodes are decommisioned
+                    self._cluster = rados.Rados(conffile="/etc/ceph/ceph.conf")
+                logger.debug(f"cluster state: {self._cluster.state}")
+                if self._cluster.state != "connected":
+                    # this can throw (auth failed, etc.)
+                    logger.debug("Connecting to cluster")
+                    self._cluster.connect()
+                    has_aquarium_pool = "aquarium" in self._cluster.list_pools()
+                    if not has_aquarium_pool:
+                        # consider setting pg_num 1 as with device_health_metrics pool
+                        # also need to do the equivalent of
+                        # `ceph osd pool application enable aquarium aquarium`
+                        # (see devicehealth mgr module)
+                        self._cluster.create_pool("aquarium")
+                    self._ioctx = self._cluster.open_ioctx("aquarium")
+                    # This actually seems to be safe (doesn't trash existing omap
+                    # data if present, which is neat)
+                    self._ioctx.write_full(
+                        "kvstore",
+                        "# aquarium kv store is in this object's omap\n".encode(
+                            "utf-8"
+                        ),
+                    )
+                    # At this point, iff it's a new pool, new object, etc.
+                    # we need to push everything from our local cache to
+                    # the omap on our kvstore, to populate it with whatever
+                    # may have been set pre-bootstrap.
+                    # Probably want to lock the object here too, in case we get two
```

Comment:
Would assert_version work instead of locking the object?
e.g `self._ioctx.assert_version(1)` before the set_omap below

Reply:
Maybe? I didn't know about that piece of functionality. Thanks for the tip :-)

Reply:
Yep, that works :-)